### ISSUE TITLE
MulticlassClassification.Trainers.StochasticDualCoordinateAscent

### ISSUE THREAD

**Issue by climosgit** [dotnet/machinelearning#2313]
### System information

- **Windows 7 Enterprise**:
- **.NET Framework 4.6.1**: 

### Issue

- **What did you do? I'm trying to predict an error classification based on error message. I have three classifications. Dismiss, Immediate and Observe. Trained, evaluated, saved and now consuming the model file. I'm using the MulticlassClassification.Trainers.StochasticDualCoordinateAscent trainer.**
- **What happened? It works fine for errors that exist in the training data. Predicting very well. However, for new error that comes in, it appears to be grabbing the classification with largest population. In my case, it's the Dismiss.**
- **What did you expect? I really wanted to classify the new error to flag as Immediate as at this time it is unknown and needs to be investigated. Perhaps I am using the wrong trainer and instead of StochasticDualCoordinateAscent, I should be using another trainer?**

### Source code / logs

mlContext = new MLContext(seed: 0);
trainErrorsEnumerable = TrainErrors(trainingViewVersion);
trainErrorsDataView = mlContext.CreateStreamingDataView(trainErrorsEnumerable);
testErrorsEnumerable = TestErrors();
testErrorsDataView = mlContext.CreateStreamingDataView(testErrorsEnumerable);

dataProcessPipeline = mlContext.Transforms.Conversion.MapValueToKey("Label")
.Append(mlContext.Transforms.Categorical.OneHotEncoding("Message", "MessageEncoded"))

modelBuilder = new Common.ModelBuilder<Error, Models.ErrorPrediction>(mlContext, dataProcessPipeline);
trainer = mlContext.MulticlassClassification.Trainers.StochasticDualCoordinateAscent(labelColumn: "Label",featureColumn: "Features");
modelBuilder.AddTrainer(trainer);
modelBuilder.Train(trainErrorsDataView);
var metrics = modelBuilder.EvaluateMultiClassClassificationModel(testErrorsDataView, "Label");
modelBuilder.SaveModelAsFile(modelFilePath);
var modelScorer = new Common.ModelScorer<Error, Models.ErrorPrediction>(mlContext);
modelScorer.LoadModelFromZipFile(modelFilePath);
var predictionScores = modelScorer.PredictSingle(error);

**Comment by Ivanidzo4ka** [comment]
Am I correctly understand what you train you data on 3 different labels, and for this 3 labels it predicts really well. But if you start to run tests on data which suppose to be marked with label which wasn't exposed during training it would produce awful results?

If that the case, this is how generally any our trainer works. It learns to predict results for 3 labels (in your case), and if you want to get good results for 4th label, you need to introduce such examples into your training dataset, so we trainer can be aware what 4th label exist.

**Comment by Ivanidzo4ka** [comment]
So I see following. You have some data with Label as string, and some Message.
You convert Message into MessageEncoded column which is vector of floats where only one value is 1, and everything else is 0.
I hope you have Features vector but I don't know it's size.
First you don't append MessageEncoded to Features vector, so that information just got ignored.
Second, What you describe sound like overfitting problem. How big is your dataset and how big is your Features column?

**Comment by climosgit** [comment]
Thanks for your quick reply Ivan. I am only utilizing two columns from my training data.
The Label (e.g. "Dismiss", "Immediate", "Observe") and the Message (e.g. error messages).
Both fields return values in string data type.
Ok.. does the new code make sense to you? Instead of using OneHotEncoding, doing the conversion instead to DataKind.Text. Otherwise, can you provide a sample code snippet?
dataProcessPipeline = mlContext.Transforms.Conversion.MapValueToKey("Label")
.Append(mlContext.Transforms.Conversion.ConvertType("Message", "MessageEncoded", DataKind.Text))
.Append(mlContext.Transforms.Concatenate("Features", "MessageConverted"));

Regarding the size of my training dataset, here's an aggregated result for our reference.
select Label, count(*) as messagecount
from trainingview
group by Label

Dismiss 47923
Immediate 2040
Observe 37

So say for example:
Any new message comes in. This message was not included in the trainingview because it's a new message. Current code always predicts a "Dismiss". Is there a way to train the model to always predict new messages to be "Immediate"? Also, I am going to test the code above tomorrow. I'll let you know how it goes. Thanks for your help!

**Comment by Ivanidzo4ka** [comment]
Let's start from basics. 
Our pipelines work with IDataView objects, which you can think of as some kind of SQL View: you have columns, each column has name, certain type, and you have rows with data.

You start with StreamingDataView which by default takes all public fields/properties and convert them into columns. If you have string property with name Label, your DataView will have column Label with type string (well internally we cast it to ReadOnlyMemory<char> but that's details).
You also have column Message with type of string.

All our multiclass learners required two columns: 

One need to be label column (it doesn't have to have name Label, it just literal label) which need to be Key type (you can think of it as Enum + it has dimensionality, it's hard to train if you don't know how many different types of labels are present).

Another column should be features column (it doesn't have to name Features, we just find them (Label and Features) convenient to have this names). This column should be vector of floats, since all our algorithms only know how to work on vector of floats.

In order to convert them to desired types we have transformers. They basically take one dataview and produce another dataview. Some of them require pass over data to learn internal information, some of them don't.

So you have Label column, and as I said before you need to have label column with Key type. This is why you have `MapValueToKey` transform. That transform iterate over all data, look for all possible values, and creates mapping (by default we map by occurrence in data) so internally we have something like this: Dismiss- 1, Immediate - 2 Observe -3. And since you didn't specify output column name, we just replace Label column with string values to Label column with Key type.

Next thing you have is Message which is string, and we want Features column with vector of floats instead. So we need somehow transform string into vector of floats. 

You initially tried to use OneHotEncoding transformer. It works following way: it goes through data, and remembers all unique values in training dataset. After that it creates array of floats with size equal to amount of unique values, fill it with zeros, and put 1 once into respective position in array for certain value. For example if you had values "A","B","C" for "A" it gonna be 1,0,0; for "B" it's gonna be 0,1,0; and for "C" it's gonna be 0,0,1. If in your test data we will encounter "D" array for it gonna be 0,0,0 since "D" wasn't present in training data. It mostly used if your data is some kind of enumeration/categories for something.

It's a good transform, but you gonna have awful model out of it, because your feature space is quite limited, and all you train is some kind of mapping between message and label. You don't need machine learning for that, you can write simple handcrafted switch statement which would work better than ML model.

If your Message is not one word, but actual sentence(or set of sentences, more data is better!), I would suggest you to use Text transform.  It would break your text into tokens, clean some stopwords, produce ngrams out of words and chars calculate TF/IDF, run normalization and create you nice beautiful array of floats.

```
dataProcessPipeline = mlContext.Transforms.Conversion.MapValueToKey("Label")
.Append(mlContext.Transforms.Text.FeaturizeText("Message", "Features")));
````

Regarding distribution of your data, as you can see it's quite skewed towards one class. It's quite easy to just return Dismiss all the time, and you will get good metrics. As option I would suggest to introduce Weight column into your data class, and play with it values. You can look on Weight column as data multiplication column, in a way, if it's 1 for row - no changes needed, if it's two then learner will treat this row as it present twice in dataset, if it's 100 as it present 100 times. This way you can set weight for Dismiss labels to 1, Immidiate to 20 and Observe to 1200 to bring some balance in your data distribution.  (I used integer values for simplification purpose, you can use real numbers as well).

**Comment by climosgit** [comment]
Thanks for the detailed explanation Ivan! That helps a lot, cleared up some confusions too!
This simply fixed the prediction results for new messages not exposed during training.
            dataProcessPipeline = mlContext.Transforms.Conversion.MapValueToKey("Label")
                .Append(mlContext.Transforms.Text.FeaturizeText("Message", "Features"));
Obviously I was incorrectly appending the message encoded to the features vector too.  :-)
I haven't utilized the Weight column yet but soon we will.
Thanks again!
Problem solved!
